library/Zend/ServiceMananger/ServiceManager.php

```diff
@@ -0,0 +1,246 @@
+     */
+    protected $sharedByDefault = true;
+
+    /**
+     * Allow to activate/deactivate shared per service name
+     *
+     * Example configuration:
+     *
+     * 'shared' => [
+     *     MyService::class => true, // will be shared, even if "sharedByDefault" is false
+     *     MyOtherService::class => false // won't be shared, even if "sharedByDefault" is true
+     * ]
+     *
+     * @var array
+     */
+    protected $shared = [];
+
+    /**
+     * @var ServiceLocatorInterface
+     */
+    protected $creationContext;
+
+    /**
+     * @param array $config
+     */
+    public function __construct(array $config)
+    {
+        $this->creationContext = $this;
+        $this->configure($config);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This is a highly performance sensitive method, do not modify if you have not benchmarked it carefully
+     */
+    public function get($name, array $options = [])
+    {
+        // We start by checking if the service is cached (this is the fastest method)
+        if (isset($this->services[$name])) {
+            return $this->services[$name];
+        }
+
+        // Let's create the service by fetching the factory
+        $factory = $this->getFactory($name);
+
+        try {
+            if (!isset($this->delegators[$name])) {
+                $object = $factory($this->creationContext, $name, $options);
+            } else {
+                $object = $this->createDelegatorFromFactory($name, $options);
+            }
+        } catch (Exception $exception) {
+            throw new ServiceNotCreatedException(sprintf(
+                'Service with name "%s" could not be created',
+                $exception->getCode(),
+                $exception
+            ));
+        }
+
+        if (($this->sharedByDefault && !isset($this->shared[$name]))
+            || (isset($this->shared[$name]) && $this->shared[$name])) {
+            $this->services[$name] = $object;
+        }
+
+        foreach ($this->initializers as $initializer) {
+            $initializer($this->creationContext, $object);
+        }
+
+        return $object;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function has($name, $checkAbstractFactories = false)
+    {
+        $found = isset($this->services[$name]) || isset($this->factories[$name]);
+
+        if ($found || !$checkAbstractFactories) {
+            return $found;
+        }
+
+        // Check abstract factories
+        foreach ($this->abstractFactories as $abstractFactory) {
+            if ($abstractFactory->canCreateServiceWithName($name)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Get a factory for the given service name
+     *
+     * @param  string $name
+     * @return callable
+     * @throws ServiceNotFoundException
+     */
+    protected function getFactory($name)
+    {
+        $factory = isset($this->factories[$name]) ? $this->factories[$name] : null;
```

Comment:
Should we have a second cache for factories, that would map a factory string name to an instance? I'm considering this because of the new InvokableFactory, you will have as many instantiated InvokableFactory than you have of object using it. Having said that, the cost of an object like a factory should be pretty low, and you don't have thousands of invokables, anyway.